### PR TITLE
Draft of CCLR for computing polarizability

### PR DESCRIPTION
**Pull request by jattakumi**
## Description
This PR implements the "asymmetric" CCLR function to compute polarizability within PyCC.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] Constructed the "asymmetric" linear response function within 'ccresponse.py' including the first-order lambda amplitudes.
  - [x]  A test case (test_036_lr.py) was used to compute the polarizability of H20/aug-cc-pvdz.

## Status
- [x] Ready to go

### CODE REVIEW

File: pycc/ccresponse.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
 """
 ccresponse.py: CC Response Functions
 """
+from opt_einsum import contract
```
Comment (lothian) [inline pycc/ccresponse.py:4]: This should be handled using something like `contract = self.ccwfn.contract` within the function that needs it.  This gives access to either the `opt_einsum` or PyTorch version.  See `solve_right()` for an example.

File: pycc/ccresponse.py [hunk 2]
```diff
@@ -283,6 +286,8 @@ def linresp(self, A, B, omega, e_conv=1e-13, r_conv=1e-13, maxiter=200, max_diis
                 pertkey = B + "_" + self.cart[axis]
                 X_key = pertkey + "_" + f"{omega:0.6f}"
                 print("Solving right-hand perturbed wave function for %s:" % (X_key))
+                #X_2[pertkey] = self.solve_right(self.pertbar[pertkey], omega, e_conv, r_conv, maxiter, max_diis, start_diis)
```
Comment (lothian) [inline pycc/ccresponse.py:289]: Remove extraneous code and commented-out lines.

File: pycc/ccresponse.py [hunk 3]
```diff
@@ -291,7 +296,58 @@ def linresp(self, A, B, omega, e_conv=1e-13, r_conv=1e-13, maxiter=200, max_diis
...
 
-    def solve_right(self, pertbar, omega, e_conv=1e-13, r_conv=1e-13, maxiter=200, max_diis=8, start_diis=1):
+
+    def linresp_asym(self, pertkey_a, X1_B, X2_B, Y1_B, Y2_B):
+
+        # Defining the l1 and l2
+        l1 = self.cclambda.l1
+        l2 = self.cclambda.l2
+
+        # Please refer to eqn 78 of [Crawford:xxxx].
```
Comment (lothian) [inline pycc/ccresponse.py:306]: Complete the reference.

File: pycc/ccresponse.py [hunk 4]
```diff
@@ -291,7 +296,58 @@ def linresp(self, A, B, omega, e_conv=1e-13, r_conv=1e-13, maxiter=200, max_diis
...
+        # Writing H(1)(omega) = B, T(1)(omega) = X, L(1)(omega) = y
+        # <<A;B>> = <0|Y(B) * A_bar|0> + <0| (1 + L(0))[A_bar, X(B)}|0>
+        #                 polar1                polar2
+        polar1 = 0
+        polar2 = 0
+        pertbar_A = self.pertbar[pertkey_a]
+        Avvoo = pertbar_A.Avvoo.swapaxes(0, 2).swapaxes(1, 3)
```
Comment (lothian) [inline pycc/ccresponse.py:313]: Why isn't this done within `perbar()`?

File: pycc/tests/test_036_lr.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    hbar = pycc.cchbar(cc)
+    cclambda = pycc.cclambda(cc, hbar)
+    lecc = cclambda.solve_lambda(e_conv, r_conv)
+    density = pycc.ccdensity(cc, cclambda)
+
+    resp = pycc.ccresponse(density)
+
+    omega1 = 0.0656
+
+    # Creating dictionaries
```
Comment (lothian) [inline pycc/tests/test_036_lr.py:38]: I don't want the user interface to be so complicated to obtain the polarizability, but we can stick with this for now and make those changes on another PR.

File: pycc/ccresponse.py [hunk 6]
```diff
@@ -391,14 +504,310 @@ def r_X2(self, pertbar, omega):
 
         return r_X2
 
+    def in_Y1(self, pertbar, X1, X2):
```
Comment (lothian) [inline pycc/ccresponse.py:507]: I don't understand the `in_Y1()`/`r_Y1()` and `in_Y2()/r_Y2()` combinations.  Why can't we just have `r_Y1()` and `r_Y2()` in analogy to the corresponding right-hand perturbed wave function residuals?

Update: Wait, maybe I do.  Are these the components of the `r_Y1` and `r_Y2` residuals that aren't dependent on `Y1` and `Y2` so you only want to compute them once?

File: pycc/ccresponse.py [hunk 7]
```diff
@@ -391,14 +504,310 @@ def r_X2(self, pertbar, omega):
...
+        o = self.ccwfn.o
+        v = self.ccwfn.v
+        l1 = self.cclambda.l1
+        l2 = self.cclambda.l2
+        cclambda = self.cclambda
+        t2 = self.ccwfn.t2
+        hbar = self.hbar
+        L = self.H.L
+
+        # Inhomogenous terms appearing in Y1 equations
```
Comment (lothian) [inline pycc/ccresponse.py:518]: I would move this up and elaborate so that it's clear to others right away what's going on in these functions.

Comment (jattakumi) [reply pycc/ccresponse.py:518]: I have moved it up and elaborated on the solve left function.

File: pycc/ccresponse.py (hunk 6 above)
Comment (jattakumi) [reply pycc/ccresponse.py:507]: Yes

File: pycc/tests/test_036_lr.py (hunk 5 above)
Comment (jattakumi) [reply pycc/tests/test_036_lr.py:38]: Okay

File: pycc/ccresponse.py (hunk 3 above)
Comment (jattakumi) [reply pycc/ccresponse.py:306]: Reference completed

File: pycc/ccresponse.py (hunk 2 above)
Comment (jattakumi) [reply pycc/ccresponse.py:289]: Okay

File: pycc/ccresponse.py (hunk 1 above)
Comment (jattakumi) [reply pycc/ccresponse.py:4]: Okay
